pingcap/tidb-operator
# Support to suspend components

### What problem does this PR solve?


### What is changed and how does it work?

For apis pkg:
* add an interface `Cluster` to abstract `TidbCluster` and `DMCluster`, and add some helper functions in `ComponentStatus`.
* add a field `SuspendAction` in `TidbCluster` and `DMCluster` for component suspending.

For manager pkg:
* Add a controller `Suspender` to support suspending any component.
* Call `suspender.SuspendComponent()` in the beginning of every component's reconciliation.

### Code changes


### Tests



* Suspend all components
1. Deploy a cluster;
4. Set `spec.suspendAction.suspendStatefulSet` to `true`;
5. Check all components are suspended in order;
6. Set `spec.suspendAction.suspendStatefulSet` to `false`;
7. Check all component is created, and the cluster is normal;

* For every component
1. Deploy a cluster;
2. Set `spec.<component>.suspendAction.suspendStatefulSet` to `true`;
3. Check the components are suspended in order;
4. Set `spec.<component>.suspendAction.suspendStatefulSet` to `false`;
5. Check the component is created, and the cluster is normal;


### Side effects


### Related changes


### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Support to suspend any component
```
